### PR TITLE
[contracts] Add ProxyFactory with CREATE2 functionality

### PR DESCRIPTION
**Pull request by snario**
Added a ProxyFactory with create2.

### CODE REVIEW

**Comment by snario** [comment]
Added a test that passes but waiting for @marekkirejczyk to push `ethereum-waffle@2.0.6` (https://github.com/EthWorks/Waffle/pull/81).

Not sure why `create2` isn't actually adding the contract to the chain...

**Comment by marekkirejczyk** [comment]
FYI Waffle 2.0.6 released

**Comment by snario** [comment]
Same error still as of today

**Comment by snario** [comment]
This seems to be the only issue on the underlying tech which is blocking us: https://github.com/ethereumjs/ethereumjs-vm/issues/356

**Comment by snario** [comment]
Seems to be working for contracts test, this fixes #1184

File: packages/contracts/contracts/proxies/ProxyFactory.sol [hunk 1]
```diff
@@ -18,11 +17,57 @@ contract ProxyFactory {
...
+  function proxyRuntimeCode() public pure returns (bytes memory) {
+    return type(Proxy).runtimeCode;
+  }
+
+  /// @dev Allows to retrieve the creation code used for the Proxy deployment. With this it is easily possible to calculate predicted address.
+  function proxyCreationCode() public pure returns (bytes memory) {
+    return type(Proxy).creationCode;
+  }
+
+  /// @dev Allows to create new proxy contact and execute a message call to the new proxy within one transaction.
```
Comment (cf19drofxots) [inline packages/contracts/contracts/proxies/ProxyFactory.sol]: Should be more specific about this function wrt a contract creation vs. address determination.

File: packages/contracts/contracts/proxies/ProxyFactory.sol [hunk 2]
```diff
@@ -18,11 +17,57 @@ contract ProxyFactory {
...
+  /// @param saltNonce Nonce that will be used to generate the salt to calculate the address of the new proxy contract.
+  function createProxyWithNonce(
+    address _mastercopy,
+    bytes memory initializer,
+    uint256 saltNonce
+  )
+    public
+    returns (Proxy proxy)
+  {
+    // If the initializer changes the proxy address should change too. Hashing the initializer data is cheaper than just concatinating it
```
Comment (cf19drofxots) [inline packages/contracts/contracts/proxies/ProxyFactory.sol]: `changes, `

File: packages/contracts/test/create2.spec.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  function create2(
+    initcode: string,
+    saltNonce: number = 0,
+    initializer: string = "0x"
+  ) {
+    return getAddress(
+      solidityKeccak256(
+        ["bytes1", "address", "uint256", "bytes32"],
+        [
+          "0xff",
```
Comment (cf19drofxots) [inline packages/contracts/test/create2.spec.ts]: Link to solidity docs wrt how this function signature works.

**Pull request merged**